pubsub/kafka/kafka.go

```diff
@@ -0,0 +1,175 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+// ------------------------------------------------------------
+
+package kafka
+
+import (
+	"context"
+	"os"
+	"os/signal"
+	"strings"
+	"sync"
+	"syscall"
+
+	"github.com/Shopify/sarama"
+	log "github.com/Sirupsen/logrus"
+	"github.com/dapr/components-contrib/pubsub"
+)
+
+// Kafka allows reading/writing to a Kafka consumer group
+type Kafka struct {
+	producer      sarama.SyncProducer
+	topics        []string
+	consumerGroup string
+	brokers       []string
+}
+
+type kafkaMetadata struct {
+	Brokers       []string `json:"brokers"`
+	Topics        []string `json:"topics"`
+	PublishTopic  string   `json:"publishTopic"`
+	ConsumerGroup string   `json:"consumerGroup"`
+}
+
+type consumer struct {
+	ready    chan bool
+	callback func(msg *pubsub.NewMessage) error
+}
+
+func (consumer *consumer) Setup(sarama.ConsumerGroupSession) error {
+	close(consumer.ready)
+	return nil
+}
+
+func (consumer *consumer) ConsumeClaim(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	for message := range claim.Messages() {
+		if consumer.callback != nil {
+			err := consumer.callback(&pubsub.NewMessage{
+				Topic: claim.Topic(),
+				Data:  message.Value,
+			})
+			if err == nil {
+				session.MarkMessage(message, "")
+			}
+		}
+	}
+	return nil
+}
+
+func (consumer *consumer) Cleanup(sarama.ConsumerGroupSession) error {
+	return nil
+}
+
+// NewKafka returns a new kafka pubsub instance
+func NewKafka() pubsub.PubSub {
+	return &Kafka{}
+}
+
+// Init does metadata parsing and connection establishment
+func (k *Kafka) Init(metadata pubsub.Metadata) error {
+	meta, err := k.getKafkaMetadata(metadata)
+	if err != nil {
+		return err
+	}
+
+	p, err := k.getSyncProducer(meta)
+	if err != nil {
+		return err
+	}
+
+	k.brokers = meta.Brokers
+	k.producer = p
+	k.topics = meta.Topics
+	k.consumerGroup = meta.ConsumerGroup
+	return nil
+}
+
+// Publish message to Kafka cluster
+func (k *Kafka) Publish(req *pubsub.PublishRequest) error {
+	_, _, err := k.producer.SendMessage(&sarama.ProducerMessage{
+		Topic: req.Topic,
+		Value: sarama.ByteEncoder(req.Data),
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Subscribe to topic in the Kafka cluser
+func (k *Kafka) Subscribe(req pubsub.SubscribeRequest, handler func(msg *pubsub.NewMessage) error) error {
+
+	config := sarama.NewConfig()
+	config.Version = sarama.V1_0_0_0
+
+	cs := consumer{
+		callback: handler,
+		ready:    make(chan bool),
+	}
+
+	client, err := sarama.NewConsumerGroup(k.brokers, k.consumerGroup, config)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			if err = client.Consume(ctx, k.topics, &cs); err != nil {
```

Comment:
this should be from `req.Topic`

Reply:
Looks like the Subscription is a 1 to 1 to a Topic, right?